hack/cluster: download cfssl if not present

dcbw opened the pull request:
hack/local-up-cluster.sh uses cfssl to generate certificates and
will exit it cfssl is not already installed.  But other cluster-up
mechanisms (GCE) that generate certs just download cfssl if not
present.  Make local-up-cluster.sh do that too so users don't have
to bother installing it from somewhere.

dcbw committed:
hack/cluster: download cfssl if not present

hack/local-up-cluster.sh uses cfssl to generate certificates and
will exit it cfssl is not already installed.  But other cluster-up
mechanisms (GCE) that generate certs just download cfssl if not
present.  Make local-up-cluster.sh do that too.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/42748)
<!-- Reviewable:end -->

dcbw commented:
cc @jbeda

dcbw commented:
Needs some fixes, reworking...

dcbw commented:
@k8s-bot cvm gce e2e test this

dcbw committed:
hack/cluster: consolidate cluster/ utils to hack/lib/util.sh

Per Clayton's suggestion, move stuff from cluster/lib/util.sh to
hack/lib/util.sh.  Also consolidate ensure-temp-dir and use the
hack/lib/util.sh implementation rather than cluster/common.sh.

dcbw commented:
Seems to be failure to copy files from nodes on test completion?  Also, fluentd is getting oom-killed a lot.

@k8s-bot gci gce e2e test this issue 41647

dcbw commented:
@smarterclayton this what you had in mind for consolidating cluster/ -> hack/ ?

smarterclayton commented:
Yes, much better.

cblecker commented on hack/lib/util.sh:
```diff
@@ -542,12 +542,14 @@ kube::util::download_file() {
     if ! command -v cfssl &>/dev/null || ! command -v cfssljson &>/dev/null; then
-      echo "Failed to successfully run 'cfssl', please verify that cfssl and cfssljson are in \$PATH."
-      echo "Hint: export PATH=\$PATH:\$GOPATH/bin; go get -u github.com/cloudflare/cfssl/cmd/..."
-      exit 1
```
As this is a test function, is there a fail condition? what if download-cfssl doesn't work?

dcbw replied:
Will add some bits to make failure here more clear.

dcbw commented:
@cblecker look better now?

cblecker commented:
@dcbw The language around the error is great.

Is there a reason why we need two different functions for cfssl (`kube::util::test_cfssl_installed` vs `kube::util::download-cfssl`)? Could these two functions be unified? If there is a technical reason why we need two, then I'm good with this. If there's a chance at simplifying the code into one function that handles both testing if present, and installing if not, that would be preferable I'd think (similar to `kube::util::ensure_godep_version`).

Which brings up another thought.. are we pinning to a specific version of cfssl? If the function `kube::util::test_cfssl_installed` as written is called, does it matter if the version of cfssl differs from the 1.2 version you're downloading?

dcbw commented:
> Is there a reason why we need two different functions for cfssl (kube::util::test_cfssl_installed vs kube::util::download-cfssl)? Could these two functions be unified? If there is a technical reason why we need two, then I'm good with this. If there's a chance at simplifying the code into one function that handles both testing if present, and installing if not, that would be preferable I'd think (similar to kube::util::ensure_godep_version).

@cblecker I'm just moving code around, but I think they could be unified.  The only part that cares whether cfssl is installed is hack/local-up-cluster.sh.  The other users of cfssl are all GCE-related and they presumably never have cfssl installed and will always fail the test and download.  So I will combine.

> Which brings up another thought.. are we pinning to a specific version of cfssl? If the function kube::util::test_cfssl_installed as written is called, does it matter if the version of cfssl differs from the 1.2 version you're downloading?

As far as I can tell, no version is getting pinned.  Obviously at this point it doesn't matter, but if cfssl changes options in the future it will.  @deads2k any idea if this matters?  You're the original author of the download-cfssl bits in 8360bc1a9fd844794c753ea1c7977bf96331c03c

deads2k commented:
> As far as I can tell, no version is getting pinned. Obviously at this point it doesn't matter, but if cfssl changes options in the future it will. @deads2k any idea if this matters?

I think its unlikely that they'll break compatibility for their tool.

dcbw commented:
@cblecker is that sufficient to answer your question?

cblecker commented:
@dcbw: Yup!

dcbw commented:
@lavalamp good to go then?

cblecker commented:
@dcbw The version pinning question was answered, but were we still going to look at possibly combining the two redundant functions?

dcbw commented:
@cblecker PTAL, thanks!  Consolidated the two functions into one kube::util::ensure-cfssl.

cblecker commented:
@dcbw awesome! will review this tomorrow :)
/assign

cblecker commented:
@k8s-bot gci gce e2e test this
@k8s-bot kubemark e2e test this
@k8s-bot gce etcd3 e2e test this
@k8s-bot non-cri e2e test this
@k8s-bot cvm gce e2e test this

cblecker commented:
@dcbw Tests are broken.
```
I0331 11:50:01.024] Failed to successfully run 'cfssl' from //google-cloud-sdk/bin:/google-cloud-sdk/bin:/go/bin:/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin; downloading instead...
W0331 11:50:01.124] Created [https://www.googleapis.com/compute/v1/projects/k8s-jkns-pr-gce/global/firewalls/e2e-gce-agent-pr-24-0-default-internal-master].
I0331 11:50:01.225] NAME                                           NETWORK                SRC_RANGES  RULES                                       SRC_TAGS  TARGET_TAGS
I0331 11:50:01.225] e2e-gce-agent-pr-24-0-default-internal-master  e2e-gce-agent-pr-24-0  10.0.0.0/8  tcp:1-2379,tcp:2382-65535,udp:1-65535,icmp            e2e-gce-agent-pr-24-0-master
W0331 11:50:02.966] mkdir: cannot create directory '/tmp/tmp.Ode6hcgzIO/cfssl/cfssl': File exists
W0331 11:50:02.974] 2017/03/31 11:50:02 util.go:98: Step './hack/e2e-internal/e2e-up.sh' finished in 3m4.62680886s
```

dcbw commented:
@cblecker should be fixed now. PTAL thanks!

cblecker commented:
@k8s-bot non-cri e2e test this

dcbw commented:
@k8s-bot non-cri e2e test this

cblecker commented on hack/lib/util.sh:
```diff
@@ -694,6 +680,95 @@ EOF
+  fi
+
+  mkdir -p "${cfssldir}"
+  pushd "${cfssldir}"
```
Stylistically, when we use pushd in these scripts, we indent the lines between pushd and popd by two so it's clear which lines are affected by the pushd. Additionally we redirect the output from pushd to /dev/null.

cblecker commented on hack/lib/util.sh:
```diff
@@ -694,6 +680,95 @@ EOF
+  CFSSL_BIN="${cfssldir}/cfssl"
+  CFSSLJSON_BIN="${cfssldir}/cfssljson"
+
+  popd
```
Output from popd is typically redirected to /dev/null

cblecker commented on hack/lib/util.sh:
```diff
@@ -694,6 +680,95 @@ EOF
+      exit 2
+  esac
+
+  if [ "${success}" == "0" ]; then
```
Instead of using a success variable here (or in any of the spots it's set above), could we use `if [[ -x ${CFSSL_BIN} -x ${CFSSLJSON_BIN} ]]`, and move this if to the end of the file? This would test not only that files downloaded, but that the permissions were able to be set correctly, and the variable is good to go (as it's the output of the function).

cblecker commented:
@k8s-bot non-cri e2e test this

The failure around PD doesn't seem related, but this flake seems persistent. Going to run one more time. If those failures persist, there may be something we're missing that the test is catching (I've run other PRs through this test successfully, so it's odd that this failure is persisting if it's actually a flake)

cblecker commented:
Okay, so now it passes. Definitely a flake. Odd that it failed 4 times in a row.

cblecker commented:
@k8s-bot gci gce e2e test this
/lgtm
/release-note-none

This looks fantastic! Thanks @dcbw :)

dcbw commented:
@lavalamp mind taking a quick look?  Thanks!

cblecker commented:
ping @lavalamp

dcbw commented:
Not sure why submit queue says Jenkins Bazel Build is not green, cause it is.  But perhaps kicking it will help...

@k8s-bot bazel test this

fejta commented:
testing 1234

lavalamp commented:
We gotta get a bash expert in the hack/ OWNERs file. I'm taking everyone's word that this is good to go...

/approve

cblecker commented:
@ixdy Hmm.. that's no good. Is there a test somewhere for the get.k8s.io script that we should have noticed or used? Also, to confirm my understanding, this doesn't impact the release-1.6 branch, as the rest of the `cluster/` scripts are a part of the kubernetes tarball itself that's downloaded by `cluster/get-kube.sh`?

Options I see:
- Copy the functions needed for cluster (or the whole file) from `hack/lib/util.sh` back to `cluster/` and redirect the source. Would fix the issue quickly, but then code is in two places.
- Add `hack/lib/` to the main tarball. I don't know how complicated this is, but that directory would add <100KB uncompressed. This would prevent the need for code duplication.

@smarterclayton @dcbw @ixdy What do you think? I prefer the second, but I don't know enough about the release packaging code to know how complicated this would be.

cblecker commented:
I've opened #44687 which I believe should fix this with the second option. If we decide to go with the first, I'll close that PR.

ixdy commented:
There is a test, implicitly, in that our CI uses `get-kube.sh`, but it also extracts the test tarball, so it masked this bug. :(
